lib/Transforms/YosysOptimizer/YosysOptimizer.cpp

```diff
@@ -0,0 +1,122 @@
+#include "mlir/include/mlir/IR/IRMapping.h"             // from @llvm-project
+#include "mlir/include/mlir/Pass/PassManager.h"         // from @llvm-project
+#include "mlir/include/mlir/Transforms/Passes.h"        // from @llvm-project
+
+namespace mlir {
+namespace heir {
+
+// $0: verilog filename
+// $1: function name
+// $2: yosys runfiles
+// $3: abc path
+// $4: abc fast option -fast
+constexpr std::string_view kYosysTemplate = R"(
+read_verilog {0};
+hierarchy -check -top \{1};
+proc; memory;
+techmap -map {2}/techmap.v; opt;
+abc -exe {3} -lut 3 {4};
+opt_clean -purge;
+rename -hide */c:*; rename -enumerate */c:*;
+techmap -map {2}/map_lut_to_lut3.v; opt_clean -purge;
+hierarchy -generate * o:Y i:*; opt; opt_clean -purge;
+clean;
+)";
+
+struct YosysOptimizer
+    : public PassWrapper<YosysOptimizer, mlir::OperationPass<mlir::ModuleOp>> {
+  YosysOptimizer(std::string runfiles, std::string abcPath, bool abcFast)
+      : runfiles(runfiles), abcPath(abcPath), abcFast(abcFast) {}
+
+  void runOnOperation() override;
+
+  llvm::StringRef getArgument() const override { return "yosys-optimizer"; }
+
+  void getDependentDialects(mlir::DialectRegistry &registry) const override {
+    registry.insert<mlir::arith::ArithDialect, mlir::heir::comb::CombDialect>();
+  }
+
+ private:
+  // Path to a directory containing yosys techlibs.
+  std::string runfiles;
+  // Path to ABC binary.
+  std::string abcPath;
+
+  bool abcFast;
+};
+
+// Globally optimize an MLIR module.
+void YosysOptimizer::runOnOperation() {
+  getOperation().walk([&](func::FuncOp op) {
+    // Translate function to Verilog. Translation will fail if the func
+    // contains unsupported operations.
+    // FIXME: Directly convert MLIR to Yosys' AST instead of using Verilog.
+    char *filename = tmpnam(NULL);
+    std::error_code EC;
+    llvm::raw_fd_ostream of(filename, EC);
+    if (failed(translateToVerilog(op, of)) || EC) {
+      return WalkResult::interrupt();
+    }
+    of.close();
+
+    // Invoke Yosys to translate to a combinational circuit and optimize.
+    Yosys::yosys_setup();
+    Yosys::log_error_stderr = true;
+    Yosys::run_pass(llvm::formatv(kYosysTemplate.data(), filename,
+                                  op.getSymName(), runfiles, abcPath,
+                                  abcFast ? "-fast" : ""));
+
+    // Translate to MLIR and insert into the func
+    std::stringstream cellOrder;
+    Yosys::log_streams.push_back(&cellOrder);
+    Yosys::run_pass("torder -stop * P*;");
+    Yosys::log_streams.clear();
+    auto topologicalOrder = getTopologicalOrder(cellOrder);
+
+    // Insert the optimized MLIR.
+    LUTImporter lutImporter = LUTImporter(&getContext());
+    Yosys::RTLIL::Design *design = Yosys::yosys_get_design();
+    func::FuncOp func =
+        lutImporter.importModule(design->top_module(), topologicalOrder);
+    op.getBody().takeBody(func.getBody());
+
+    func.dump();
```

Comment:
I like the idea of printing, but let's hide it behind an LLVM_DEBUG. See, e.g., this little PR I upstreamed to MLIR just adding debug messages: https://github.com/llvm/llvm-project/pull/69476

It will get printed when you run with `--debug`, and maybe let's dress it up a little with `llvm::outs() << "Converted & optimized func op via yosys. Input func:\n" << op << "\n\nOutput func:\n" << func << std::endl`

Reply:
hah! this was leftover from some debugging, but this is great to know. This will be useful as this pass adds the local optimization option.